### PR TITLE
Adding parallel capabilities to Galerkin solvers

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

Adding parallel methods for Galerkin. Both global and local devices. The parallel implementation is a so called "slab" decomposition, where spectral arrays are sharded along the first axis and physical arrays are sharded along the second.

## Types of Changes

- [ ] Bug fix
- [x] New feature
- [x] Performance improvement
- [x] Documentation update
- [ ] Refactor / cleanup
- [ ] Build / CI
- [ ] Other (describe):

## Description of Changes

Add device mesh and sharding to each TensorProductSpace instance. The device mesh is distributing one single axis over all devices, both local and global:
```python
_spmd_mesh = Mesh(jax.devices(), ("k",))
```
Two shardings are described 
```python
_spectral_sharding = NamedSharding(_spmd_mesh, P("k", None))
_physical_sharding = NamedSharding(_spmd_mesh, P(None, "k")) 
```
A complete parallel solver (here simulating 2 local devices) is shown below. It can be run with mpirun.
```python
# Solve Poisson's equation in 2D
# ruff: noqa: E402
import os
import socket
import sys

import jax

jax.config.update("jax_enable_x64", True)
jax.config.update("jax_num_cpu_devices", 2)

def initialize_distributed():
    import jax.distributed as jdist

    try:
        from mpi4py import MPI
    except Exception:
        return

    comm = MPI.COMM_WORLD
    world = comm.Get_size()
    rank = comm.Get_rank()

    if world == 1:
        return

    # Rank 0 chooses a free TCP port on localhost
    if rank == 0:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.bind(("127.0.0.1", 0))
        host, port = s.getsockname()
        s.close()
        coord = f"{host}:{port}"
    else:
        coord = None
    coord = comm.bcast(coord, root=0)

    # Initialize jax.distributed
    jdist.initialize(
        coordinator_address=coord,
        num_processes=world,
        process_id=rank,
    )


initialize_distributed()

import jax.numpy as jnp
import sympy as sp

from jaxfun.galerkin.arguments import TestFunction, TrialFunction, x, y
from jaxfun.galerkin.Fourier import Fourier
from jaxfun.galerkin.functionspace import FunctionSpace
from jaxfun.galerkin.inner import inner
from jaxfun.galerkin.Legendre import Legendre
from jaxfun.galerkin.tensorproductspace import TensorProduct
from jaxfun.operators import Div, Grad
from jaxfun.utils.common import lambdify, n, ulp

ue = (1 - y**2) * (sp.cos(2 * x)) * sp.exp(sp.cos(sp.pi * y))

M, N = 80, 128
bcs = {"left": {"D": ue.subs(y, -1)}, "right": {"D": ue.subs(y, 1)}}
D = FunctionSpace(M, Legendre, bcs, scaling=n + 1, name="D", fun_str="psi")
F = FunctionSpace(N, Fourier, name="F", fun_str="E")
T = TensorProduct(F, D, name="T")
v = TestFunction(T, name="v")
u = TrialFunction(T, name="u")

# Method of manufactured solution
x, y = T.system.base_scalars()
ue = T.system.expr_psi_to_base_scalar(ue)

# Serial assembly, but returns sharded array b
A, b = inner(v * Div(Grad(u)) - v * Div(Grad(ue)), sparse=True)

# Parallel solve across all devices
uh = A.solve(b)

# Parallel backward
uj = T.backward(uh)

# Create replicated mesh and evaluate exact solution on it
xj = T.mesh(broadcast=True)
uej = lambdify((x, y), ue)(*xj)

# compute error. Note that uh is sharded whereas uej is replicated.
error = jnp.linalg.norm(uj - uej) / M

# Test forward parallel transform
un = T.forward(uj)
error_fwd = jnp.linalg.norm(un - uh) / M

# Test parallel evaluation
xj = jnp.array([[0.5, 0.5], [0.6, 0.6], [0.7, 0.7]])
f = T.evaluate(xj, uh)
error_eval = jnp.linalg.norm(f - lambdify((x, y), ue)(*xj.T)) / xj.shape[1]
print("Evaluation error =", error_eval)
```
## Breaking Changes

- [ ] Yes (describe below)
- [x] No

If yes, describe what breaks and migration guidance.

## Checklist

- [ ] Tests added or updated
- [x] Documentation updated (README, examples, docstrings)
- [x] Added type hints
- [x] Linting & formatting pass locally (`pre-commit run --all-files`)
- [x] All new and existing tests pass (`uv run pytest`)
- [ ] Coverage does not decrease
- [x] Git history is clean (squash/fixup before merge)

## Additional notes

Not sure what to do with parallel testing??

### CODE REVIEW

File: src/jaxfun/la/tpmatrix.py [hunk 1]
```diff
@@ -603,78 +604,242 @@ class TPMatricesWavenumberSolver:
...
+            # ---- Fast batched path: one vmapped XLA call for all wavenumbers ----
+            # Avoids the O(n_F) Python loop of per-wavenumber B.lu_factor() calls.
+            n_F, _n_diags, n_P_local = B_data_batch.shape
+            _dtype = B_data_batch.dtype
+            p = max((-o for o in poly_offsets if o < 0), default=0)
+            q = max((o for o in poly_offsets if o > 0), default=0)
+            center = p
+            bw = p + q + 1
+            all_L_offsets: tuple[int, ...] = tuple(o for o in poly_offsets if o < 0)
+            all_U_offsets: tuple[int, ...] = tuple(o for o in poly_offsets if o >= 0)
```
[the remark this answers is not in the document]
Comment (mikaem) [reply src/jaxfun/la/tpmatrix.py]: Fixed downstream

File: src/jaxfun/la/tpmatrix.py [hunk 2]
```diff
@@ -603,78 +604,242 @@ class TPMatricesWavenumberSolver:
...
+        else:
+            # ---- Legacy path: per-wavenumber Python loop ----------------------
+            # Kept for backward compatibility when B_data_batch is not provided.
+            assert B_matrices is not None, (
+                "Either B_data_batch+poly_offsets or B_matrices must be provided."
+            )
+            n_P_local = B_matrices[0].shape[0]
+            _all_offsets = sorted({off for B in B_matrices for off in B.offsets})
+            all_L_offsets = tuple(o for o in _all_offsets if o < 0)
+            all_U_offsets = tuple(o for o in _all_offsets if o >= 0)
```
[the remark this answers is not in the document]
Comment (mikaem) [reply src/jaxfun/la/tpmatrix.py]: Fixed downstream

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 3]
```diff
@@ -280,6 +288,69 @@ def evaluate(self, x: Array, c: Array, use_einsum: bool) -> Array:
...
+            dim = len(self)
+            T = self.basespaces
+
+            # Per-axis evaluation matrices — fully replicated.
+            # eval_basis_functions is @jit_vmap'd: 1-D input (n_pts,) → (n_pts, N).
+            # x has shape (n_pts, dim); x[:, i] gives axis-i coordinates.
+            C = [
+                T[i].eval_basis_functions(T[i].map_reference_domain(x[:, i]))
+                for i in range(dim)
+            ]
```
[the remark this answers is not in the document]
Comment (mikaem) [reply src/jaxfun/galerkin/tensorproductspace.py:327]: Docstring is wrong. Fixed downstream.

File: src/jaxfun/galerkin/inner.py [hunk 4]
```diff
@@ -513,6 +514,10 @@ def process_results(
 
         return ares1D, bresults
 
+    assert not isinstance(test_space, OrthogonalSpace)
+    if len(jax.devices()) > 1:
+        bresults: Array = jax.device_put(bresults, test_space._spectral_sharding)
```
[the remark this answers is not in the document]
Comment (mikaem) [reply src/jaxfun/galerkin/inner.py]: _test_space._spectral_sharding is None only if len(jax.devices()) == 1, so no need for the getattr. Adding len(bresults) > 0 downstream.

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 5]
```diff
@@ -476,6 +720,20 @@ def __init__(
...
+            None
+            if len(jax.devices()) == 1
+            else NamedSharding(self._spmd_mesh, P(None, "k", None))
+        )
+        # Sharding of arrays in physical space.
+        self._physical_sharding: NamedSharding | None = (
+            None
+            if len(jax.devices()) == 1
+            else NamedSharding(self._spmd_mesh, P(None, None, "k", None))
+        )
```
[the remark this answers is not in the document]
Comment (mikaem) [reply src/jaxfun/galerkin/tensorproductspace.py:741]: The shardings are used in inner. Keeping them.
